From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: reito <reito@chromium.org>
Date: Sat, 26 Apr 2025 22:58:30 +0800
Subject: feat: import external texture

Prevent call CloseHandle on external DXGI, as it is a external handle, and we doesn't have ownership. We can't DuplicateHandle either, because Windows needs a HANDLE to source process which we don't have.

diff --git a/ui/gfx/gpu_memory_buffer.cc b/ui/gfx/gpu_memory_buffer.cc
index 76b25a3b26a79af0f74213015ff4c8a85a2d2a16..9ba86df734b523a933c5b3678c7866c126f25b4b 100644
--- a/ui/gfx/gpu_memory_buffer.cc
+++ b/ui/gfx/gpu_memory_buffer.cc
@@ -52,14 +52,19 @@ DXGIHandle::DXGIHandle(base::win::ScopedHandle buffer_handle,
   DCHECK(buffer_handle_.is_valid());
 }
 
+DXGIHandle::DXGIHandle(HANDLE external_handle)
+    : external_handle_(external_handle) {}
+
 DXGIHandle::DXGIHandle(DXGIHandle&&) = default;
 DXGIHandle& DXGIHandle::operator=(DXGIHandle&&) = default;
 
 bool DXGIHandle::IsValid() const {
-  return buffer_handle_.is_valid();
+  return buffer_handle_.is_valid() || external_handle_.has_value();
 }
 
 DXGIHandle DXGIHandle::Clone() const {
+  DCHECK(!external_handle_.has_value());
+
   DXGIHandle handle;
   if (buffer_handle_.is_valid()) {
     handle.buffer_handle_ = CloneDXGIHandle(buffer_handle_.Get());
@@ -78,6 +83,8 @@ DXGIHandle DXGIHandle::CloneWithRegion(
 }
 
 base::win::ScopedHandle DXGIHandle::TakeBufferHandle() {
+  DCHECK(!external_handle_.has_value());
+
   DCHECK(buffer_handle_.is_valid());
   return std::move(buffer_handle_);
 }
diff --git a/ui/gfx/gpu_memory_buffer.h b/ui/gfx/gpu_memory_buffer.h
index 3f907fb57941649b08b0fe326aa46465dca9c360..0cfeaa2e9fa03f54df6efe3af05d6d119252f75a 100644
--- a/ui/gfx/gpu_memory_buffer.h
+++ b/ui/gfx/gpu_memory_buffer.h
@@ -106,6 +106,10 @@ class COMPONENT_EXPORT(GFX) DXGIHandle {
              const DXGIHandleToken& token,
              base::UnsafeSharedMemoryRegion region);
 
+  // Typically only used by IPC deserialization. `buffer_handle` must be valid,
+  // but `region` may be invalid.
+  DXGIHandle(HANDLE external_handle);
+
   DXGIHandle(DXGIHandle&&);
   DXGIHandle& operator=(DXGIHandle&&);
 
@@ -126,9 +130,14 @@ class COMPONENT_EXPORT(GFX) DXGIHandle {
   // `region().IsValid()` is false.
   DXGIHandle CloneWithRegion(base::UnsafeSharedMemoryRegion region) const;
 
-  HANDLE buffer_handle() const { return buffer_handle_.Get(); }
-  base::win::ScopedHandle TakeBufferHandle();
+  HANDLE buffer_handle() const {
+    if (external_handle_.has_value())
+      return external_handle_.value();
 
+    return buffer_handle_.Get();
+  }
+
+  base::win::ScopedHandle TakeBufferHandle();
   const DXGIHandleToken& token() const { return token_; }
 
   const base::UnsafeSharedMemoryRegion& region() const { return region_; }
@@ -140,6 +149,8 @@ class COMPONENT_EXPORT(GFX) DXGIHandle {
   base::win::ScopedHandle buffer_handle_;
   DXGIHandleToken token_;
   base::UnsafeSharedMemoryRegion region_;
+
+  std::optional<HANDLE> external_handle_;
 };
 #endif
 
